### ISSUE TITLE
Data preprocessing for tabular took too long

### ISSUE THREAD

**Issue by mli** [awslabs/autogluon#591]
I'm using autogluon to train a [kaggle competition](https://www.kaggle.com/c/ashrae-energy-prediction/data), but it took too long for autogluon to prepare the data, which may due to too many examples. 

Here are the codes to reproduce this issue. 

First download dataset

```bash
kaggle competitions download -c ashrae-energy-prediction -p ../input/
cd ../input/; unzip ashrae-energy-prediction.zip -d ashrae-energy-prediction
```

then load and merge data

```python
import pandas as pd 

building = pd.read_csv('../input/ashrae-energy-prediction/building_metadata.csv')
weather_train = pd.read_csv('../input/ashrae-energy-prediction/weather_train.csv')
weather_test = pd.read_csv('../input/ashrae-energy-prediction/weather_test.csv')
train = pd.read_csv('../input/ashrae-energy-prediction/train.csv')
test = pd.read_csv('../input/ashrae-energy-prediction/test.csv')

train = train.merge(building, on='building_id', how='left')
test = test.merge(building, on='building_id', how='left')

train = train.merge(weather_train, on=['site_id', 'timestamp'], how='left')
test = test.merge(weather_test, on=['site_id', 'timestamp'], how='left')
```

next use autogluon to train

```python
from autogluon import TabularPrediction as task

model = task.fit(train_data=task.Dataset(train),#.drop(columns=drop_columns)), 
                 label='meter_reading', output_directory='autogluon')
```

then the progress hangs (>1h) after the following log. 

```bash
Beginning AutoGluon training ...
AutoGluon will save models to autogluon/
AutoGluon Version:  0.0.12
Train Data Rows:    20216100
Train Data Columns: 16
Preprocessing data ...
Here are the first 10 unique label values in your data:  [0.0, 53.2397, 43.0013, 52.4206, 23.3036, 0.3746, 175.18400000000003, 91.2653, 80.93, 86.2283]
AutoGluon infers your prediction problem is: regression  (because dtype of label-column == float and many unique label-values observed).
If this is wrong, please specify `problem_type` argument in fit() instead (You may specify problem_type as one of: ['binary', 'multiclass', 'regression'])
```

**Comment by Innixma** [comment]
Thanks for providing the code example for reproducing. We have not tried a dataset with this many rows before on AutoGluon (20M), so it may be that some of the preprocessing code is struggling to scale to this size or that memory was not sufficient.

Hopefully this should become easier to debug and fix once the feature generator refactor (#584) is merged.

**Comment by Innixma** [comment]
The performance issue of data pre-processing has been resolved via numerous optimizations in v0.2 and v0.3 that have overall sped-up performance on large datasets by >1000x since v0.1.

Running this code on v0.3 pre-release on a 2015 13-inch dual core Mac Laptop, AutoGluon finishes pre-processing all 20 million rows in under 3 minutes.

Code:
```
import pandas as pd

path_prefix = 'datasets/ashrae-energy-prediction'

building = pd.read_csv(f'{path_prefix}/building_metadata.csv')
weather_train = pd.read_csv(f'{path_prefix}/weather_train.csv')
weather_test = pd.read_csv(f'{path_prefix}/weather_test.csv')
train = pd.read_csv(f'{path_prefix}/train.csv')
test = pd.read_csv(f'{path_prefix}/test.csv')

train = train.merge(building, on='building_id', how='left')
test = test.merge(building, on='building_id', how='left')

train = train.merge(weather_train, on=['site_id', 'timestamp'], how='left')
test = test.merge(weather_test, on=['site_id', 'timestamp'], how='left')

from autogluon.tabular import TabularPredictor

model = TabularPredictor(label='meter_reading').fit(train_data=train)

```

Logs:
```
Beginning AutoGluon training ...
AutoGluon will save models to "AutogluonModels/ag-20210814_233408/"
AutoGluon Version:  0.3.0b20210814
Train Data Rows:    20216100
Train Data Columns: 15
Preprocessing data ...
AutoGluon infers your prediction problem is: 'regression' (because dtype of label-column == float and many unique label-values observed).
	Label info (max, min, mean, stddev): (21904700.0, 0.0, 2117.12108, 153235.62499)
	If 'regression' is not the correct problem_type, please manually specify the problem_type argument in fit() (You may specify problem_type as one of: ['binary', 'multiclass', 'regression'])
Using Feature Generators to preprocess the data ...
Fitting AutoMLPipelineFeatureGenerator...
	Available Memory:                    12324.63 MB
	Train Data (Original)  Memory Usage: 5039.92 MB (40.9% of available memory)
	Warning: Data size prior to feature transformation consumes 40.9% of available memory. Consider increasing memory or subsampling the data to avoid instability.
	Inferring data type of each feature based on column values. Set feature_metadata_in to manually specify special dtypes of the features.
	Stage 1 Generators:
		Fitting AsTypeFeatureGenerator...
	Stage 2 Generators:
		Fitting FillNaFeatureGenerator...
	Stage 3 Generators:
		Fitting IdentityFeatureGenerator...
		Fitting CategoryFeatureGenerator...
			Fitting CategoryMemoryMinimizeFeatureGenerator...
		Fitting DatetimeFeatureGenerator...
	Stage 4 Generators:
		Fitting DropUniqueFeatureGenerator...
	Types of features in original data (raw dtype, special dtypes):
		('float', [])                      : 9 | ['year_built', 'floor_count', 'air_temperature', 'cloud_coverage', 'dew_temperature', ...]
		('int', [])                        : 4 | ['building_id', 'meter', 'site_id', 'square_feet']
		('object', [])                     : 1 | ['primary_use']
		('object', ['datetime_as_object']) : 1 | ['timestamp']
	Types of features in processed data (raw dtype, special dtypes):
		('category', [])             : 1 | ['primary_use']
		('float', [])                : 9 | ['year_built', 'floor_count', 'air_temperature', 'cloud_coverage', 'dew_temperature', ...]
		('int', [])                  : 4 | ['building_id', 'meter', 'site_id', 'square_feet']
		('int', ['datetime_as_int']) : 1 | ['timestamp']
	142.9s = Fit runtime
	15 features in original data used to generate 15 features in processed data.
	Train Data (Processed) Memory Usage: 2284.42 MB (12.2% of available memory)
Data preprocessing and feature engineering runtime = 171.06s ...
AutoGluon will gauge predictive performance using evaluation metric: 'root_mean_squared_error'
	To change this, specify the eval_metric argument of fit()
Automatically generating train/validation split with holdout_frac=0.01, Train Rows: 20013939, Val Rows: 202161

...
```
